### PR TITLE
[GH-262] Merge request approval subscription

### PR DESCRIPTION
**Pull request by MatthewDorner**
#### Summary
Add DM and Channel notifications for Merge Request `approved` and `unapproved` webhook actions.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-gitlab/issues/282

### CODE REVIEW

File: server/webhook/merge_request.go [hunk 1]
```diff
@@ -83,6 +87,10 @@ func (w *webhook) handleChannelMergeRequest(ctx context.Context, event *gitlab.M
 		message = fmt.Sprintf("[%s] Merge request [!%v %s](%s) was closed by [%s](%s)", repo.PathWithNamespace, pr.IID, pr.Title, pr.URL, senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername))
 	case actionReopen:
 		message = fmt.Sprintf("[%s] Merge request [!%v %s](%s) was reopened by [%s](%s)", repo.PathWithNamespace, pr.IID, pr.Title, pr.URL, senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername))
+	case actionApproved:
+		message = fmt.Sprintf("[%s] Merge request [!%v %s](%s) was approved by [%s](%s)", repo.PathWithNamespace, pr.IID, pr.Title, pr.URL, senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername))
+	case actionUnapproved:
+		message = fmt.Sprintf("[%s] Merge request [!%v %s](%s) was unapproved by [%s](%s)", repo.PathWithNamespace, pr.IID, pr.Title, pr.URL, senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername))
```
Comment (spirosoik) [inline server/webhook/merge_request.go:93]: ditto. something like `needs changes`

**Review by spirosoik — CHANGES_REQUESTED** [review]
a few changes

**Comment by MatthewDorner** [comment]
Not sure why the CircleCI checks are timing out.

**Comment by MatthewDorner** [comment]
Changed the messages, anything else to do?

File: server/webhook/merge_request.go [hunk 2]
```diff
@@ -35,6 +35,10 @@ func (w *webhook) handleDMMergeRequest(event *gitlab.MergeEvent) ([]*HandleWebho
 		case actionUpdate:
 			// TODO not enough check (opened/update) to say assigned to you...
 			message = fmt.Sprintf("[%s](%s) assigned you to merge request [%s!%v](%s)", senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername), event.ObjectAttributes.Target.PathWithNamespace, event.ObjectAttributes.IID, event.ObjectAttributes.URL)
+		case actionApproved:
+			message = fmt.Sprintf("[%s](%s) approved your merge request [%s!%v](%s)", senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername), event.ObjectAttributes.Target.PathWithNamespace, event.ObjectAttributes.IID, event.ObjectAttributes.URL)
+		case actionUnapproved:
+			message = fmt.Sprintf("[%s](%s) requested changes to your merge request [%s!%v](%s)", senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername), event.ObjectAttributes.Target.PathWithNamespace, event.ObjectAttributes.IID, event.ObjectAttributes.URL)
```
Comment (mickmister) [inline server/webhook/merge_request.go:41]: GitHub's convention is to use `#` here, not sure about GitLab. I see other references to `!` here though

```suggestion
			message = fmt.Sprintf("[%s](%s) requested changes to your merge request [%s#%v](%s)", senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername), event.ObjectAttributes.Target.PathWithNamespace, event.ObjectAttributes.IID, event.ObjectAttributes.URL)
```

File: server/webhook/merge_request.go [hunk 3]
```diff
@@ -83,6 +87,10 @@ func (w *webhook) handleChannelMergeRequest(ctx context.Context, event *gitlab.M
 		message = fmt.Sprintf("[%s] Merge request [!%v %s](%s) was closed by [%s](%s)", repo.PathWithNamespace, pr.IID, pr.Title, pr.URL, senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername))
 	case actionReopen:
 		message = fmt.Sprintf("[%s] Merge request [!%v %s](%s) was reopened by [%s](%s)", repo.PathWithNamespace, pr.IID, pr.Title, pr.URL, senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername))
+	case actionApproved:
+		message = fmt.Sprintf("[%s] Merge request [!%v %s](%s) was approved by [%s](%s)", repo.PathWithNamespace, pr.IID, pr.Title, pr.URL, senderGitlabUsername, w.gitlabRetreiver.GetUserURL(senderGitlabUsername))
```
Comment (mickmister) [inline server/webhook/merge_request.go:91]: (Not part of this PR), but ideally we use the `sender`'s connected Mattermost account to display their Mattermost username instead of their GitLab username

**Review by mickmister — APPROVED** [review]

**Comment by mickmister** [comment]
> Changed the messages, anything else to do?

@MatthewDorner Make sure you either mention someone when asking a question or re-request review from the person if that's the context of the question

File: server/webhook/merge_request.go (hunk 3 above)
Comment (MatthewDorner) [reply server/webhook/merge_request.go:91]: Couldn't the sender be somebody who hasn't connected their account?

File: server/webhook/merge_request.go (hunk 2 above)
Comment (MatthewDorner) [reply server/webhook/merge_request.go:41]: GitLab does use `!` (https://docs.gitlab.com/ee/user/markdown.html) but the MR references in the Channel messages are formatted differently from the DM ones. Looks like the DM ones are doing it the correct way.

I can quickly fix it here.

Comment (MatthewDorner) [reply server/webhook/merge_request.go:41]: In this screenshot, the second set of messages would be after my proposed changes where you can see the DM and Channel messages now format the MR reference in the same way. But I'd go even further and suggest that the sentence structure should also be the same instead of one being "X approved MR" and the other "MR was approved by". There's a bunch of little stuff like that I'm seeing but probably should be separate PRs.

2022_07_24_0lg_Kleki

Comment (MatthewDorner) [reply server/webhook/merge_request.go:41]: @mickmister should I also mention the reviewer in replies to review threads like this one? It's been a while since I used GitHub...

Anyway I don't think I should commit the screenshotted change as part of this PR. I'd rather review all the different types of webhook message templates and propose fixes in a later PR, but I do think there are multiple things that should be fixed.

But the use of `!` itself is correct for GitLab MRs.

**Review by spirosoik — APPROVED** [review]
thanks for the contribution

File: server/webhook/merge_request.go (hunk 3 above)
Comment (mickmister) [reply server/webhook/merge_request.go:91]: Yes, but we should do our best to use the connected account. If they aren't connected, then we should fall back to the default functionality

File: Makefile [hunk 4]
```diff
@@ -88,7 +88,7 @@ endif
 ## Ensures NPM dependencies are installed without having to run this all the time.
 webapp/node_modules: $(wildcard webapp/package.json)
 ifneq ($(HAS_WEBAPP),)
-	cd webapp && $(NPM) install
+	cd webapp && $(NPM) install --verbose
```
Comment (hanzei) [inline Makefile:91]: That seems like an unrelated change

Comment (MatthewDorner) [reply Makefile:91]: hmm, that was committed by @mickmister in  [##30c35d2](https://github.com/mattermost/mattermost-plugin-gitlab/pull/307/commits/30c35d2c93b0ad2849ea40c754bcf20789141797) I don't remember anything about this PR that would be related to that.

Comment (mickmister) [reply Makefile:91]: This was put here due to something related to CI, but that has since been fixed. We should make sure we're not re-introducing this change here. To be clear `--verbose` should not be used

Comment (MatthewDorner) [reply Makefile]: OK, I reverted it

**Comment by hanzei** [comment]
@DHaussermann Gentle reminder to review this PR

**Review by DHaussermann — APPROVED** [review]
Tested and passed 
- Created a subscription that includes `merge_request_comments` 
- Confirmed that both Approval and Revoke Approval now send webhook events for this feature.
- A DM is sent to the user 

LGTM!

Thanks @MatthewDorner for implementing this! 🎉
Sorry for the delay in getting this approved.

**Comment by hanzei** [comment]
@MatthewDorner Would you please merge `master` into your branch?

**Comment by MatthewDorner** [comment]
@hanzei done

**Comment by hanzei** [comment]
Thank you for the contribution @MatthewDorner :+1:
